File: backend/app/models/watch.py
from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import Boolean, DateTime, ForeignKey, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db import Base
from app.models.organization import Organization

# What kind of source this is. Free-form column constrained to the known set by
# the Pydantic schema — same choice as Interaction.kind and Deal.stage.
#
# The two that matter most are *not* companies: a job board with a saved search
# and a tender portal are sources you sweep, not parties you have a relationship
# with. A careers page is the special case where the source does have an
# organization behind it.
WATCH_KINDS = ("job_board", "careers_page", "tender_portal", "other")

# What a sweep turned up. "nothing" is the normal, valuable answer — it is what
# makes a year of diligence on a quiet portal provable.
CHECK_OUTCOMES = ("nothing", "found")


class Watch(Base):
    """A source swept on a cadence: a job board, a careers page, a tender portal.

    The intake end of the pipeline. Everything from Deal onward records
    opportunities that already exist; this is where they come from.
    """

    __tablename__ = "watches"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    user_id: Mapped[UUID] = mapped_column(ForeignKey("user.id", ondelete="CASCADE"), index=True)
    name: Mapped[str]
    url: Mapped[str]
    kind: Mapped[str] = mapped_column(String, default="other", server_default="other", index=True)

    # The saved search in words: keywords, CPV codes, region — the filters the
    # URL encodes. A query string is unreadable six months later.
    query_note: Mapped[str | None]
    notes: Mapped[str | None]

    # Set for a careers page, null for a portal. Deliberately not enforced per
    # kind: finding a company you have not filed yet is normal, and refusing the
    # watch until you do is backwards. SET NULL, like every other link here.
    organization_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("organizations.id", ondelete="SET NULL"), index=True
    )

    # The cadence, reusing the rule set and arithmetic from app/recurrence.py —
    # the same closed dropdown, the same month-end clamping, and the same
    # "next due is computed from the check, not the missed slot" rule. What it
    # does *not* reuse is Task: one task per watch would bury the real
    # follow-ups under twenty identical "check X" rows.
    recurrence_rule: Mapped[str]
    recurrence_interval: Mapped[int] = mapped_column(default=1, server_default="1")

    # Stamped together whenever a check is logged, and nowhere else, so they
    # cannot drift from the WatchCheck history they summarise. Stored rather
    # than derived because the list has to sort and page by "most overdue" in
    # SQL, and month-clamping arithmetic does not belong in a query.
    last_checked_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    next_due_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), index=True)

    # Pause a source without deleting it and losing its history.
    active: Mapped[bool] = mapped_column(Boolean, default=True, server_default="true")

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    organization: Mapped[Organization | None] = relationship(lazy="selectin")

    @property
    def organization_name(self) -> str | None:
        """Denormalised for reads, so a watch list needs no second request."""
        return self.organization.name if self.organization is not None else None


class WatchCheck(Base):
    """One sweep of one watch. Append-only.

    Not a `last_checked_at` column overwritten each time, for the reason
    recurring tasks already commit to — "did I actually check in March?" — and
    one more: a source that has produced nothing in a year has to look
    different from a fresh one, which a single timestamp cannot show.
    """

    __tablename__ = "watch_checks"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    user_id: Mapped[UUID] = mapped_column(ForeignKey("user.id", ondelete="CASCADE"), index=True)
    # CASCADE: the log is part of the watch, not an independent record.
    watch_id: Mapped[UUID] = mapped_column(ForeignKey("watches.id", ondelete="CASCADE"), index=True)
    checked_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), index=True)
    # One of CHECK_OUTCOMES.
    outcome: Mapped[str] = mapped_column(String, default="nothing", server_default="nothing")
    note: Mapped[str | None]

    # What this find turned into, so the source of a win stays traceable back to
    # the sweep that found it. SET NULL, not CASCADE: deleting the deal must not
    # erase the record of having found it.
    created_deal_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("deals.id", ondelete="SET NULL"), index=True
    )
    created_task_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("tasks.id", ondelete="SET NULL"), index=True
    )

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    @property
    def found(self) -> bool:
        return self.outcome == "found"

File: frontend/lib/models/watch.dart
import '../core/date_time_text.dart';
import 'task.dart' show recurrenceUnits;

/// What kind of source this is.
///
/// The two that matter most are **not** companies: a job board with a saved
/// search and a tender portal are sources you sweep, not parties you have a
/// relationship with. A careers page is the special case where the source does
/// have an organization behind it.
enum WatchKind {
  jobBoard('job_board', 'Job board', 'Job boards'),
  careersPage('careers_page', 'Careers page', 'Careers pages'),
  tenderPortal('tender_portal', 'Tender portal', 'Tender portals'),
  other('other', 'Other', 'Other');

  const WatchKind(this.wire, this.label, this.plural);

  /// The value the API uses. Kept separate from [label] so renaming what the
  /// operator reads never changes what is stored.
  final String wire;
  final String label;
  final String plural;

  /// Unknown kinds fall back rather than throwing: a backend that learns a new
  /// kind must not blank the whole screen on an older build.
  static WatchKind fromWire(String? value) => WatchKind.values.firstWhere(
        (kind) => kind.wire == value,
        orElse: () => WatchKind.other,
      );
}

/// What a sweep turned up. "nothing" is the normal, valuable answer.
enum CheckOutcome {
  nothing('nothing', 'Nothing found'),
  found('found', 'Found something');

  const CheckOutcome(this.wire, this.label);

  final String wire;
  final String label;

  static CheckOutcome fromWire(String? value) => CheckOutcome.values.firstWhere(
        (outcome) => outcome.wire == value,
        orElse: () => CheckOutcome.nothing,
      );
}

/// A source swept on a cadence: a job board, a careers page, a tender portal.
class Watch {
  const Watch({
    required this.id,
    required this.name,
    required this.url,
    this.kind = WatchKind.other,
    this.queryNote,
    this.notes,
    this.organizationId,
    this.organizationName,
    required this.recurrenceRule,
    this.recurrenceInterval = 1,
    this.lastCheckedAt,
    required this.nextDueAt,
    this.active = true,
    this.foundCount = 0,
    this.checkCount = 0,
  });

  final String id;
  final String name;
  final String url;
  final WatchKind kind;

  /// The saved search in words: keywords, CPV codes, region. A query string is
  /// unreadable six months later.
  final String? queryNote;
  final String? notes;

  final String? organizationId;

  /// Denormalised by the API so a list row shows the company without a request
  /// per row.
  final String? organizationName;

  /// One of [recurrenceRules] — the same closed set tasks use.
  final String recurrenceRule;
  final int recurrenceInterval;

  final DateTime? lastCheckedAt;
  final DateTime nextDueAt;
  final bool active;

  /// How many sweeps ever turned something up. The question a single timestamp
  /// cannot answer: is this source worth keeping?
  final int foundCount;
  final int checkCount;

  /// Due now or overdue. A paused source is never due, whatever its date says.
  bool get isDue => active && !nextDueAt.isAfter(DateTime.now());

  /// Whole days past due; 0 when it is not overdue.
  int get daysOverdue {
    if (!isDue) return 0;
    return DateTime.now().difference(nextDueAt).inDays;
  }

  bool get neverChecked => lastCheckedAt == null;

  /// "Every week", "Every 2 months".
  String get cadenceLabel => recurrenceInterval == 1
      ? 'Every ${recurrenceUnitSingular[recurrenceRule] ?? recurrenceRule}'
      : 'Every $recurrenceInterval ${recurrenceUnits[recurrenceRule] ?? recurrenceRule}';

  /// "Never checked", "Due today", "3 days overdue", "Due 2026-09-20".
  String get dueLabel {
    if (neverChecked) return 'Never checked';
    if (!isDue) return 'Due ${formatDay(nextDueAt)}';
    final days = daysOverdue;
    if (days == 0) return 'Due today';
    return days == 1 ? '1 day overdue' : '$days days overdue';
  }

  factory Watch.fromJson(Map<String, dynamic> json) => Watch(
        id: json['id'] as String,
        name: json['name'] as String,
        url: json['url'] as String,
        kind: WatchKind.fromWire(json['kind'] as String?),
        queryNote: json['query_note'] as String?,
        notes: json['notes'] as String?,
        organizationId: json['organization_id'] as String?,
        organizationName: json['organization_name'] as String?,
        recurrenceRule: json['recurrence_rule'] as String? ?? 'weekly',
        recurrenceInterval: json['recurrence_interval'] as int? ?? 1,
        lastCheckedAt: json['last_checked_at'] == null
            ? null
            : DateTime.parse(json['last_checked_at'] as String),
        nextDueAt: DateTime.parse(json['next_due_at'] as String),
        active: json['active'] as bool? ?? true,
        foundCount: json['found_count'] as int? ?? 0,
        checkCount: json['check_count'] as int? ?? 0,
      );
}

/// The singular noun a cadence counts, keyed as `recurrenceRules`.
const recurrenceUnitSingular = {
  'daily': 'day',
  'weekly': 'week',
  'monthly': 'month',
  'yearly': 'year',
};

/// One sweep of one watch. Append-only — there is no edit.
class WatchCheck {
  const WatchCheck({
    required this.id,
    required this.watchId,
    required this.checkedAt,
    required this.outcome,
    this.note,
    this.createdDealId,
    this.createdTaskId,
  });

  final String id;
  final String watchId;
  final DateTime checkedAt;
  final CheckOutcome outcome;
  final String? note;

  /// What this find turned into, so the source of a win stays traceable. Null
  /// again once that deal or task is deleted — the record of finding it stays.
  final String? createdDealId;
  final String? createdTaskId;

  bool get found => outcome == CheckOutcome.found;

  factory WatchCheck.fromJson(Map<String, dynamic> json) => WatchCheck(
        id: json['id'] as String,
        watchId: json['watch_id'] as String,
        checkedAt: DateTime.parse(json['checked_at'] as String),
        outcome: CheckOutcome.fromWire(json['outcome'] as String?),
        note: json['note'] as String?,
        createdDealId: json['created_deal_id'] as String?,
        createdTaskId: json['created_task_id'] as String?,
      );
}

/// The response to logging a sweep: the check, plus the watch it moved on.
class WatchCheckResult {
  const WatchCheckResult({required this.check, required this.watch});

  final WatchCheck check;
  final Watch watch;

  factory WatchCheckResult.fromJson(Map<String, dynamic> json) => WatchCheckResult(
        check: WatchCheck.fromJson(json['check'] as Map<String, dynamic>),
        watch: Watch.fromJson(json['watch'] as Map<String, dynamic>),
      );
}
